value.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"sync"
+
+	promproto "github.com/prometheus/client_model/go"
+	"go.uber.org/atomic"
+)
+
+// Value is an atomic with some associated metadata. It's a building block
+// for higher-level metric types.
+type value struct {
+	atomic.Int64
+
+	meta       metadata
+	labelPairs []*promproto.LabelPair
+}
+
+func newValue(m metadata) value {
+	return value{
+		meta:       m,
+		labelPairs: m.MergeLabels(nil /* variable labels */),
+	}
+}
+
+func newDynamicValue(m metadata, variableLabels []string) value {
+	return value{
+		meta:       m,
+		labelPairs: m.MergeLabels(variableLabels),
+	}
+}
+
+func (v value) snapshot() SimpleSnapshot {
+	return SimpleSnapshot{
+		Name:   *v.meta.Name,
+		Labels: zip(v.labelPairs),
+		Value:  v.Load(),
+	}
+}
+
+// A vector is a collection of values that share the same metadata.
+type vector struct {
+	meta metadata
+
+	// The factory function creates a new metric given the vector's metadata
+	// and the variable label keys and values.
+	factory func(metadata, []string) metric
+
+	metricsMu sync.RWMutex
+	metrics   map[string]metric // key is variable label vals
+}
+
+func (vec *vector) getOrCreate(labels ...string) (metric, error) {
```

Comment:
this is an internal api, any reason to not use `[]string`?

Reply:
Additionally, this is representing the "name", "value" pair for the labels right?  Might make sense to be as explicit as possible about this in the variable name:

labelNameThenValues
labelNameAndValues
labelNameValuePairs

Or something along those lines.  It took me a little while to realize that these are not the just the names, and not just the values.

Reply:
Can definitely change to a slice. I'm open to changing the variable names, but not right now - if we're trying to get all this migrated in the next 24h, I don't want to burn time running around renaming multiple formal parameters.

Reply:
👍